###########################################################################
# This script will align the reads generated by "simulate_shotgun_reads.sh"
# It also creates an index from the bam file so that we can compute 
# coverage later on.
###########################################################################

#Author - Kobie Kirven
#Date - 10-07-2021


#File paths
DATA_PATH=../data/10-06-2021/

#Make a directory for the alignment files
mkdir ${DATA_PATH}alignments

#########################
#Align reads with bowtie2
#########################

#Align male reads to reference genome
bowtie2 --local -x ${DATA_PATH}bowtie2_index/GRCh38_latest_genomic \
-1 ${DATA_PATH}reference_genome/reads/male_10000_1.fa \
-2 ${DATA_PATH}reference_genome/reads/male_10000_2.fa \
> ../data/10-06-2021/alignments/male_10000_align.sam
samtools view -bS ../data/10-06-2021/alignments/male_10000_align.sam \
> ../data/10-06-2021/alignments/male_10000_align.bam

#Align female reads to reference genome
bowtie2 --local -x ${DATA_PATH}bowtie2_index/GRCh38_latest_genomic \
-1 ${DATA_PATH}reference_genome/reads/female_10000_1.fa \
-2 ${DATA_PATH}reference_genome/reads/female_10000_2.fa \
> ../data/10-06-2021/alignments/female_10000_align.sam
samtools view -bS ../data/10-06-2021/alignments/female_10000_align.sam \
> ../data/10-06-2021/alignments/female_10000_align.bam

#Sort the alignment files
samtools sort -m 5G ../data/10-06-2021/alignments/male_10000_align.bam \
-o ../data/10-06-2021/alignments/male_10000_align.bam

samtools sort -m 5G ../data/10-06-2021/alignments/female_10000_align.bam \
-o ../data/10-06-2021/alignments/female_10000_align.bam


#Create new alignment files containing only those reads that 
# map uniquely to the reference genome
#samtools view -F 2316 -bS ../data/10-06-2021/alignments/male_10000_align.bam  
#> ../data/10-06-2021/alignments/male_10000_align_mapuniq.bam 

#samtools view -F 2316 -bS ../data/10-06-2021/alignments/female_10000_align.bam \ 
#> ../data/10-06-2021/alignments/female_10000_align_mapuniq.bam 


# Create index files
samtools index ../data/10-06-2021/alignments/male_10000_align.bam \
../data/10-06-2021/alignments/male_10000_align_index.bam

samtools index ../data/10-06-2021/alignments/female_10000_align.bam \
../data/10-06-2021/alignments/female_10000_align_index.bam


# Generate coverage files
samtools depth ../data/10-06-2021/alignments/male_10000_align.bam \
> ../data/10-06-2021/alignments/male_10000_align_coverage.txt

samtools depth ../data/10-06-2021/alignments/female_10000_align.bam \
> ../data/10-06-2021/alignments/female_10000_align_coverage.txt

