<style>
    .para {font-size:18px}
    
</style>


<body style="margin:250;padding:0">
<h1>Lab Notebook for Davenport Lab</h1>
<h2>Author - Kobie Kirven</h2>
<h2>Penn State University</h2>


<h2>---------------------------</h2>
<h2>Metagenomic Sex Calling Pipeline: 10-06-2021</h2>
<h3>Introduction:</h3>
<h3>Methods:</h3>
<h3><em>Initially Downloading Data</em></h3>
<p class=para>
The human reference genome build GRCh38 was downloaded from NCBI using
the script "download_reference_genome.sh." Because I needed to be able 
to simulate reads from male and female reference genomes, I created artificial
diploid reference genomes for both sexes using "make_simulated_chromosomes.py."
Next, 10,000 random, paired-end sequences were generated using wgsim 
with default parameters implemented in the script, "simulate_shotgun_reads.sh."
Once the simulated reads were generated, they were aligned to the GRCh38 
reference genome using Bowtie2 with the "--local" flag. The aligned reads
were then sorted using "samtools sort." Next, the depths were calculated
for each position in each chromosome using "samtools depth." The alignment, 
sorting, and depth commands were implemented in the "align_with_bowtie2.sh" script.
</p>
<h3><em>Calculating Average Depth of Coverage</em></h3>
<p class=para>
    To determine the average depth for each chromosome, the total depth for 
each chromosome was computed by summing the depths at each position and the
total depth was divided by the length of that chromosome (implemented in 
"calculate_coverage_depth.py"). This process was repeated for data from both 
male and female mock genomes. The results of the X:Autosomal coverage ratios 
are shown in Figure 1.
</p>

<p class=para>
After looking at the resuts of the X:Autosomal, average coverage ratio plots, 
it was was apparent that something may have not gone correctly in the process
because the plots were identical. Because the X:X average coverage rato was
one for both the male and female simulated genome, I had the idea that the 
wgsim may not have worked as I expected and possibly it got confused by
having an input file that contained repeated FASTA IDs. To test this theory,
I creted a test FASTA file named "wgsim_test.fa" in which the IDs were the same, 
but the sequences were entirely different: one was only A repeats and the other
was only C repeats. Next, I used this FASTA file as a reference genome and
simulated 10 reads total with wgsim (implemented in "wgsim_validation.sh").
After looking at the output of the wgsim test, I learned that I was wrong 
and wgsim does not care about the FASTA IDs. It generates reads from each
sequence in the reference genome. 
</p>

<p class=para>
I next tried to figure out how the authors of the Zonkey pipeline calculated
their coverage. After looking around in their GitHub repo, I found on one
page that they caluculate hits as:
<q>
    Sum of SE, PE_1, and PE_2 hits. Note that supplementary
    alignments, duplicates, reads that failed QC, secondary
    alignments, and unmapped reads are ignored.
    </q>
Using this definition for hits, I wanted to see if we could get more accurate
X:autosome ratios. I used the sam files generated by the "align_with_bowtie2.sh" 
script and selected only those alignments that met the criteria of:
<ul>
  <li>Not Unmapped</li>
  <li>Mate not unmapped</li>
  <li>Not a secondary alignment</li>
  <li>Not a supplementary alignment</li>
</ul>

<p class=para>
Next, I counted the number of alignments that passed these filtering criteria
and grouped the counts by the chromosome the reads mapped to. Then, for each
chromosome, I divided the number of counts by the length of that chromosome.
Finally, I divided the normalized coverage of the X chromosome by the normalized
coverages for each of the autosomes. This entire process, using the adapted
version of the Zonkey method, was done for both male and female artifical 
genomes, and the results are plotted in Figure 2. 
</p>
<p class=para>
I next asked the question, does sequencing depth effect the X:Autosomal
ratios? To test this, I ran the same pipeline for simulating shotgun read
generation from an artifical genome, and I used the sequencing depths
100, 250, 500, 1000, 2500, 5000, and 7,500. To make the analysis more streamlined,
I combined the pipeline into one script titled "read_depth_pipeline.sh." The 
results from varying the seqencing depth are plotted in Figure 3. 
</p>

<p class=para>
The next question became, does preforming multiple runs of the read simulation 
and chromosome mapping give wildly different coverage results? To answer this 
question, I simulated 2500 reads from the male and female mock genome respectively
for 100 runs and plotted the results in Figure 4. Additionally, I ran the same
repeated simulations at a sequencing depth of 5000 reads, and the results
are shown in Figure 4. 
</p>
<h3>Results:</h3>

<img src="figures/x_to_autosome_female_10000.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_10000.png" height="300" width="300"  />
<p style="margin-bottom:50;" class=para>
<strong>Figure 1.</strong> Plots of the average coverage of X chromosome divided by average 
coverage for each chromosome.
</p>

<p class=para>
 <p>
 <img src="figures/x_to_autosome_female_10000_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_10000_zonkey.png" height="300" width="300"  />
</p>
<p style="margin-bottom:50;" class=para>
<strong>Figure 2.</strong> Plots of the normalized coverage of the X chromosome divided
by normalized coverage for each chromosome.
</p>
</p>

<p class=para>
 <p>
 <img src="figures/x_to_autosome_female_100_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_100_zonkey.png" height="300" width="300"  />
</p>
<p>
<img src="figures/x_to_autosome_female_250_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_250_zonkey.png" height="300" width="300"  />
</p>

<p>
<img src="figures/x_to_autosome_female_500_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_500_zonkey.png" height="300" width="300"  />
</p>

<p>
<img src="figures/x_to_autosome_female_1000_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_1000_zonkey.png" height="300" width="300"  />
</p>

<p>
<img src="figures/x_to_autosome_female_2500_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_2500_zonkey.png" height="300" width="300"  />
</p>

<p>
<img src="figures/x_to_autosome_female_5000_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_5000_zonkey.png" height="300" width="300"  />
</p>

<p>
<img src="figures/x_to_autosome_female_7500_zonkey.png" height="300" width="300"  />
<img src="figures/x_to_autosome_male_7500_zonkey.png" height="300" width="300"  />
</p>

<p style="margin-bottom:50;" class=para>
<strong>Figure 2.</strong> Plots of the normalized coverage of the X chromosome divided
by normalized coverage for each chromosome with varying sequencing depth.
</p>

<p>
<img src="figures/x_to_autosome_male_zonkey_2500_100_reps.png" height="300" width="300"  />
<img src="figures/x_to_autosome_female_zonkey_2500_100_reps.png" height="300" width="300"  />
</p>

<p>
<img src="figures/x_to_autosome_male_zonkey_5000_100_reps.png" height="300" width="300"  />
<img src="figures/x_to_autosome_female_zonkey_5000_100_reps.png" height="300" width="300"  />
</p>

<p style="margin-bottom:50;" class=para>
<strong>Figure 3.</strong> Box and whisker plots of normalized coverage for each autosome
divided by the normalized coverage for X for 100 runs at a sequenceing depth of
(top) 2500 and (bottom) 5000. 
</p>


<h3>Discussion:</h3>
<p class=para>
    
</p>
<h3>References:</h3>
<ol class=para>
  <li>
      Schubert, Mikkel, et al. "Zonkey: A simple, accurate and sensitive 
      pipeline to genetically identify equine F1-hybrids in archaeological 
      assemblages." Journal of Archaeological Science 78 (2017): 147-157.
</li>
</ol>


<h2>Testing SCIMS on Human Data</h2>
<h3>Introduction:</h3>
<p>
    The next question became, will scims actually work on real human data? 
    To test this, I ran the pipeline on fecal metagenomic data from the TwinsUK
    study.
</p>

<h3>Materials and Methods:</h3>
<p>
    Data was transfered by Emily Davenport to the "/gpfs/group/exd44/default/data/TwinsUK/BGImetagenomes"
    directory. Next, the data were quality filtered usign trimmomatic with a 
    sliding window of 4 with a minimum average quality of 30. After quality 
    filtering, the paired reads that passed the filtering step were used 
    as input for SCIMS. 
    
</p>

<h3>Results:</h3>
<p>
<img src="hmp_BGI36049.png" height="300" width="300"  />
<img src="hmp_BGI36050.png" height="300" width="300"  />
<img src="hmp_BGI36702.png" height="300" width="300"  />
<img src="hmp_BGI36703.png" height="300" width="300"  />
<img src="hmp_BGI37181.png" height="300" width="300"  />
<img src="hmp_BGI37182.png" height="300" width="300"  />
<img src="hmp_BGI37201.png" height="300" width="300"  />
<img src="hmp_BGI37202.png" height="300" width="300"  />
</p>

From left to right: (Row 1) BGI36049 BGI36050 BGI36702 BGI36703 (Row 2) BGI37181 BGI37182 BGI37201 
BGI37202

<h2>
<table>
  <tr>
    <th>Sample Name |</th>
    <th>True Genetic Sex |</th>
    <th>Total # of Reads Mapping</th>
  </tr>
  <tr>
    <td>36049</td>
    <td>F</td>
    <td>38285</td>
  </tr>
  <tr>
    <td>36050</td>
    <td>F</td>
    <td>345092</td>
  </tr>
  <tr>
    <td>36702</td>
    <td>F</td>
    <td>66208</td>
  </tr>
  <tr>
    <td>36703</td>
    <td>F</td>
    <td>1352</td>
  </tr>
  <tr>
    <td>37181</td>
    <td>F</td>
    <td>4905</td>
  </tr>
  <tr>
    <td>37182</td>
    <td>F</td>
    <td>4084</td>
  </tr>
  <tr>
    <td>37201</td>
    <td>F</td>
    <td>3632</td>
  </tr>
  <tr>
    <td>37202</td>
    <td>F</td>
    <td>2290</td>
  </tr>
</table>
</h2>

<h2 style="margin-top:100;">SCiMS With Single-end Data</h2>
<img src="tibet_SRR9850902.png" height="300" width="300"  />

